### PR TITLE
feat(MutantFilters): add MutantFilter to ignore BlockMutator for blocks that already contain active mutants

### LINKED ISSUE #2367: Conditional placement of mutations
**Is your feature request related to a problem? Please describe.**
Block mutations and statement mutations aren't always useful. For example
image
The above block mutation will always be killed if one of the containing mutations are also killed. Block mutations are only useful when no other mutations are placed inside the block.

The same goes for statment mutations. If some mutation is placed inside the statement, the statement removal will always be killed if the containing mutation is killed.

**Describe the solution you'd like**
I'd like these mutations only to be placed (or executed) when they don't contain any mutations.

We can do this during the placing of mutations. If we provide a context to mutators they can decide to not place a mutation if a mutation was already placed. This would work as we place mutations from the bottom of the AST first.

Or we can filter mutations after placing all mutations.

**Additional context**
Stryker.js also wants this feature so we might discuss the implementation with them as well.

### PR DESCRIPTION
**Pull request by Liam-Rougoor**
Closes #2367

### CODE REVIEW

File: src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs [hunk 1]
```diff
@@ -122,9 +122,34 @@ public void MutantFilterFactory_Creates_ExcludeLinqExpressionFilter_When_Exclude
...
+            // Arrange
+            var options = new StrykerOptions()
+            {
+                WithBaseline = true,
+                ExcludedLinqExpressions = new List<LinqExpression>
+                {
+                    LinqExpression.Distinct
+                },
+            };
+            var diffProviderMock = new Mock<IDiffProvider>(MockBehavior.Loose);
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs:142]: Change these to `MockBehaviour.Strict` please

File: src/Stryker.Core/Stryker.Core/Mutants/Mutant.cs [hunk 2]
```diff
@@ -1,4 +1,6 @@
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/Mutants/Mutant.cs:1]: Why this empty line?

File: src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/IgnoreBlockMutantFilterTests.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            // Arrange
+            var source = @"
+public void SomeMethod()
+{
+    var x = 1 + 1;
+}";
+
+            var syntaxTree = CSharpSyntaxTree.ParseText(source).GetRoot();
+            var blockNode = (BlockSyntax)syntaxTree.DescendantNodes().First(t => t is BlockSyntax);
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/IgnoreBlockMutantFilterTests.cs:46]: You could write this as:
```
syntaxTree.DescendantNodes().OfType<BlockSyntax>().First();
```
Which looks a bit better in my opinion

File: src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs (hunk 1 above)
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs:142]: Change these to `MockBehaviour.Strict` please

File: src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs [hunk 4]
```diff
@@ -122,9 +122,34 @@ public void MutantFilterFactory_Creates_ExcludeLinqExpressionFilter_When_Exclude
...
+            var gitInfoProviderMock = new Mock<IGitInfoProvider>(MockBehavior.Loose);
+            var baselineProviderMock = new Mock<IBaselineProvider>(MockBehavior.Loose);
+
+            // Act
+            var result = MutantFilterFactory.Create(options, null, diffProviderMock.Object, baselineProviderMock.Object, gitInfoProviderMock.Object);
+            var broadcastFilterResult = result as BroadcastMutantFilter;
+
+            // Assert
+            broadcastFilterResult.MutantFilters.Last().ShouldBeOfType<IgnoreBlockMutantFilter>();
+
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs:152]: Remove empty line

File: src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs [hunk 5]
```diff
@@ -122,9 +122,34 @@ public void MutantFilterFactory_Creates_ExcludeLinqExpressionFilter_When_Exclude
...
+        [Fact]
+        public void MutantFilterFactory_Creates_BlockMutantFilter_Last()
+        {
+            // Arrange
+            var options = new StrykerOptions()
+            {
+                WithBaseline = true,
+                ExcludedLinqExpressions = new List<LinqExpression>
+                {
+                    LinqExpression.Distinct
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/MutantFilterFactoryTests.cs:139]: Explain why we enable this option

File: src/Stryker.Core/Stryker.Core/MutantFilters/IgnoreBlockMutantFilter.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+        public MutantFilter Type => MutantFilter.IgnoreBlockRemoval;
+
+        public IgnoreBlockMutantFilter()
+        {
+            _inactiveStatuses = new HashSet<MutantStatus>
+            {
+                MutantStatus.Ignored,
+                MutantStatus.CompileError,
+                MutantStatus.Timeout,
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/MutantFilters/IgnoreBlockMutantFilter.cs:24]: You can remove this one

File: integrationtest/ValidationProject/ValidateStrykerResults.cs [hunk 7]
```diff
@@ -67,7 +67,7 @@ public void NetCoreWithTwoTestProjects()
 
             var report = JsonConvert.DeserializeObject<JsonReport>(strykerRunOutput);
 
-            CheckReportMutantCounts(report, total: 112, ignored: 6, survived: 10, killed: 14, timeout: 2, nocoverage: 78);
+            CheckReportMutantCounts(report, total: 112, ignored: 27, survived: 8, killed: 8, timeout: 2, nocoverage: 65);
```
Comment (richardwerkman) [inline integrationtest/ValidationProject/ValidateStrykerResults.cs:70]: Can you verify these extra ignored mutants are all block type?

**Review by richardwerkman — CHANGES_REQUESTED** [review]
Looking good, but needs some work ;)

File: integrationtest/ValidationProject/ValidateStrykerResults.cs (hunk 7 above)
Comment (Liam-Rougoor) [reply integrationtest/ValidationProject/ValidateStrykerResults.cs:70]: I checked the reports, it looks like they are indeed block mutants.

**Review by richardwerkman — APPROVED** [review]
🚀

**Pull request merged**